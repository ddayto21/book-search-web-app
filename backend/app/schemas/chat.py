from typing import List, Optional
from pydantic import BaseModel, Field
from datetime import datetime
import uuid


class ChatPart(BaseModel):
    """Represents a single part of a chat message."""

    type: str = "text"
    text: str


class ChatMessage(BaseModel):
    """Represents a single chat message."""

    id: Optional[str] = None
    role: str
    content: Optional[str] = None
    timestamp: Optional[str] = Field(
        default_factory=lambda: datetime.utcnow().isoformat()
    )

    parts: Optional[List[ChatPart]] = []


class ChatRequest(BaseModel):
    """Represents the request body for the /chat endpoint."""

    messages: List[ChatMessage]
