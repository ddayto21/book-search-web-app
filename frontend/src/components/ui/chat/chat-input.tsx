/**
 * @file chat-input.tsx - Chat input field and send button component.
 *
 * This component:
 * - Renders an input field and a send button.
 * - Handles user input and submission via form or Enter key.
 * - Updates UI based on chat status and loading state.
 */

import React from "react";
import { SendIcon } from "@components/icons/send-icon";
import { PauseIcon } from "@components/icons/pause-icon";

interface ChatInputProps {
  input: string;
  onInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onSend: (e: React.FormEvent<HTMLFormElement>) => void;
  status: string; // "ready" | "submitted" | "error"
}

/**
 * Memoized chat input component with form submission handling.
 */
export const ChatInput = React.memo(
  ({ input, onInputChange, onSend, status }: ChatInputProps) => {
    ChatInput.displayName = "ChatInput";

    const isSending = status === "submitted" || status === "streaming";
    const hasInput = input.trim() !== "";
    const isDisabled = status !== "ready";

    // Handle Enter key press to submit the form
    const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
      if (e.key === "Enter" && !isDisabled && hasInput) {
        e.preventDefault();
        const formEvent = new Event("submit", {
          bubbles: true,
        }) as unknown as React.FormEvent<HTMLFormElement>;
        onSend(formEvent); // Simulate form submit
      }
    };

    return (
      <footer className="flex items-center p-4 bg-black border-t border-gray-800">
        <input
          className="
        flex-1
        px-4 py-2
        rounded-full
        bg-gray-800 text-gray-100
        placeholder-gray-500
        border border-gray-700
        focus:outline-none
        mr-2
        transition-colors duration-300
      "
          placeholder="Type a message..."
          value={input}
          onChange={onInputChange}
          onKeyDown={handleKeyDown}
          disabled={isDisabled}
        />
        {hasInput && (
          <button
            type="submit"
            disabled={isDisabled}
            className={`
          flex items-center justify-center
          w-10 h-10
          rounded-full
          transition-colors duration-300
          ${
            isDisabled
              ? "bg-gray-700 text-gray-400 cursor-not-allowed"
              : "bg-blue-600 hover:bg-blue-700 text-white"
          }
        `}
          >
            {isSending ? <PauseIcon /> : <SendIcon />}
          </button>
        )}
      </footer>
    );
  }
);
