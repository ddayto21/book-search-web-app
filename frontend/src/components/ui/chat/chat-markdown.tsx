import React, { ElementType } from "react";
import Markdown from "markdown-to-jsx";

import hljs from "highlight.js";
import "highlight.js/styles/github-dark.css";

/**
 * Options for rendering markdown content using markdown-to-jsx.
 *
 * The `overrides` property allows us to specify custom rendering for specific
 * markdown elements. For example, we define custom components for headings,
 * lists, blockquotes, code blocks, etc., along with appropriate CSS classes
 * to match our UI design.
 *
 * @type {object}
 */
const markdownOptions = {
  overrides: {
    h1: {
      component: "h1" as ElementType,
      props: {
        className: "text-3xl font-bold mb-4 text-white",
      },
    },
    h2: {
      component: "h2" as ElementType,
      props: {
        className: "text-2xl font-bold mb-3 text-white",
      },
    },
    h3: {
      component: "h3" as ElementType,
      props: {
        className: "text-xl font-bold mb-2 text-white",
      },
    },
    ul: {
      component: "ul" as ElementType,
      props: {
        className: "list-disc pl-6 mb-2 text-gray-300",
      },
    },
    ol: {
      component: "ol" as ElementType,
      props: {
        className: "list-decimal pl-6 mb-2 text-gray-300",
      },
    },
    li: {
      component: "li" as ElementType,
      props: {
        className: "mb-1",
      },
    },
    blockquote: {
      component: "blockquote" as ElementType,
      props: {
        className: "border-l-4 pl-4 italic text-gray-400 my-4 text-gray-300",
      },
    },
    code: {
      component: "code" as ElementType,
      props: {
        className: "bg-gray-700 rounded px-1 py-0.5 text-gray-100",
      },
    },
    pre: {
      component: "pre" as ElementType,
      props: {
        className: "bg-gray-800 text-white p-4 rounded my-2 overflow-x-auto",
      },
      children: (props: { children: React.ReactNode }) => {
        const code = Array.isArray(props.children)
          ? props.children.join("")
          : props.children;
        const highlightedCode = hljs.highlightAuto(code as string).value;
        return <code className="hljs">{highlightedCode}</code>;
      },
    },
    p: {
      component: "p" as ElementType,
      props: {
        className: "mb-2 text-gray-300",
      },
    },
    a: {
      component: "a" as ElementType,
      props: {
        className: "text-blue-400 hover:text-blue-500",
      },
    },
  },
};
/**
 * ChatMarkdown component renders markdown content for chat messages.
 *
 * It uses markdown-to-jsx with custom options to transform markdown syntax into
 * React elements with specific styling for headings, lists, blockquotes, code blocks, etc.
 *
 * @param {object} props - Component properties.
 * @param {string} props.content - The markdown content to render.
 * @returns {JSX.Element} A React element that renders the formatted markdown.
 */
export const ChatMarkdown: React.FC<{ content: string }> = ({ content }) => {
  // Convert single newlines to "  \n" (two spaces before the newline).
  // This tells Markdown that each single newline is a line break.
  const forceLineBreaks = content.replace(/\n/g, "  \n");

  return <Markdown options={markdownOptions}>{forceLineBreaks}</Markdown>;
};
