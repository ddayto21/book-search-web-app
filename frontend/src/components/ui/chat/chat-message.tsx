import React from "react";
import { formatDistanceToNow } from "date-fns";
import { parseSseData } from "@utils/parse-sse-data";
import type { Message } from "@ai-sdk/react";

// Extend the default message interface
export interface ChatMessage extends Message {
  timestamp?: string;
}

export function formatContent(message: Message): string {
  if (
    message.role === "assistant" &&
    message.content.trim().startsWith("data:")
  ) {
    return parseSseData(message.content);
  }
  return message.content;
}

export function getTimeAgo(timestamp: string) {
  const dateObj = new Date(timestamp);
  return formatDistanceToNow(dateObj, { addSuffix: true });
}

interface ChatMessageProps {
  msg: ChatMessage;
}

export const ChatMessageComponent: React.FC<ChatMessageProps> = ({ msg }) => {
  const isAssistant = msg.role === "assistant";
  const content = formatContent(msg);
  const relativeTime = msg.timestamp ? getTimeAgo(msg.timestamp) : "";

  return (
    <div
      className={`flex items-start ${
        isAssistant ? "justify-start" : "justify-end"
      }`}
    >
      {/* Avatar */}
      {isAssistant ? (
        <div className="mr-2 flex-none">
          <div className="h-8 w-8 flex items-center justify-center bg-neutral-800 rounded-full text-sm font-bold">
            AI
          </div>
        </div>
      ) : (
        <div className="mr-2 flex-none">
          <div className="h-8 w-8 rounded-full bg-gradient-to-br from-blue-600 to-purple-600" />
        </div>
      )}

      {/* Message bubble */}
      <div
        className={`max-w-xs sm:max-w-md md:max-w-lg lg:max-w-xl px-4 py-3 rounded-2xl shadow ${
          isAssistant
            ? "bg-neutral-800 rounded-bl-none"
            : "bg-blue-600 rounded-br-none"
        }`}
      >
        {/* Header row: name + relative timestamp */}
        <div className="flex items-center gap-2 mb-1">
          <span className="font-semibold text-sm text-gray-200">
            {isAssistant ? "GenerativeAgent" : "You"}
          </span>
          {relativeTime && (
            <span className="text-xs text-gray-400">{relativeTime}</span>
          )}
        </div>
        <div className="text-sm leading-snug whitespace-pre-wrap text-gray-100">
          {content}
        </div>
      </div>
    </div>
  );
};
