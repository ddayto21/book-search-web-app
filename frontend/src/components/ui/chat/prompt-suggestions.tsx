import React, { useState } from "react";

interface PromptSuggestion {
  title: string;
  description?: string;
  content: string;
  icon?: string;
}

interface PromptSuggestionsProps {
  examplePrompts: PromptSuggestion[];
  onPromptClick: (prompt: string) => void;
}

export const PromptSuggestions: React.FC<PromptSuggestionsProps> = ({
  examplePrompts,
  onPromptClick,
}) => {
  // Tracks whether the suggestions are collapsed or expanded
  const [isCollapsed, setIsCollapsed] = useState(false);

  // Toggle collapse state
  function toggleCollapse() {
    setIsCollapsed((prev) => !prev);
  }

  return (
    <aside
      className="
        relative
        p-3
        bg-black/60
        backdrop-blur-sm
        transition-all
        duration-300
      "
      // If collapsed, reduce height drastically, hide overflow
      style={{
        maxHeight: isCollapsed ? "2rem" : "40%",
        overflowY: isCollapsed ? "hidden" : "auto",
      }}
    >
      <div className="flex items-center justify-between">
        <p className="text-base font-semibold text-gray-200">
          {isCollapsed ? "" : "Select a topic"}
        </p>
        <button
          onClick={toggleCollapse}
          className="
            text-gray-400
            hover:text-gray-200
            text-sm
            transition-colors
            bg-neutral-800/50
            rounded-full
            px-2 py-1
          "
          aria-label="Toggle suggestions"
        >
          {isCollapsed ? "▲" : "▼"}
        </button>
      </div>

     
     
      {!isCollapsed && (
        <div
          className="
            flex flex-wrap
            gap-2

            overflow-x-auto scrollbar-custom
            whitespace-nowrap
            h-full
          "
        >
          {examplePrompts.map((prompt, index) => (
            <button
              key={index}
              onClick={() => onPromptClick(prompt.content)}
              className="
                inline-flex items-center 
                bg-neutral-800 text-gray-100
                rounded-full px-4 py-2
                text-sm font-medium
                border border-neutral-700
                transition-colors 
                hover:bg-neutral-700
                whitespace-nowrap
              "
            >
              {prompt.icon && (
                <span className="mr-2" role="img" aria-hidden>
                  {prompt.icon}
                </span>
              )}
              {prompt.title}
            </button>
          ))}
        </div>
      )}
    </aside>
  );
};
