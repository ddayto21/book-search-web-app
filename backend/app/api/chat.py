from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
import json
import logging
import os
import sys
import uuid
from openai import OpenAI, AsyncOpenAI
from datetime import datetime
from typing import List, Optional
from dotenv import load_dotenv
from pydantic import BaseModel
from app.schemas.chat import ChatRequest, ChatMessage, ChatPart
import asyncio
from app.services.auth import verify_token
from fastapi import Depends


load_dotenv()

router = APIRouter()
logger = logging.getLogger(__name__)


# Ensure API Key is set
API_KEY = os.environ.get("DEEPSEEK_API_KEY")
if not API_KEY:
    raise ValueError(
        "Missing API Key. Set `DEEPSEEK_API_KEY` in environment variables."
    )

# client = OpenAI(api_key=API_KEY, base_url="https://api.deepseek.com/v1")
client = AsyncOpenAI(api_key=API_KEY, base_url="https://api.deepseek.com/v1")


@router.post("/chat")
async def chat(req: ChatRequest):
    system_prompt = ChatMessage(
        role="system",
        content=(
            "You are LexiGPT, a warm and knowledgeable AI librarian. You specialize in book recommendations "
            "and personalized suggestions. You speak in a conversational and engaging way, occasionally "
            "using light humor when appropriate. Always make the user feel heard and valued. "
            "Never be roboticâ€”respond naturally and empathetically."
            "Only give 3 recommendations at most."
            "If you don't have an appropriate recommendation, mention that you don't have enough information to know the answer."
        ),
        timestamp=datetime.utcnow().isoformat(),
        parts=[
            ChatPart(
                type="text",
                text="Only respond with 3 book recommendations.",
            )
        ],
    )

    # Insert system prompt at the beginning of the conversation history
    req.messages.insert(0, system_prompt)

    stream = await client.chat.completions.create(
        model="deepseek-chat",
        messages=req.messages,
        max_tokens=500,
        temperature=0.3,
        stream=True,
    )

    async def generator():
        async for chunk in stream:
            yield chunk.choices[0].delta.content or ""

    response_messages = generator()
    return StreamingResponse(response_messages, media_type="text/event-stream")


@router.get("/chat/messages")
async def get_messages(user=Depends(verify_token)):
    """Fetches messages for the authenticated user."""
    return {"messages": ["Message 1", "Message 2"], "user": user["email"]}
