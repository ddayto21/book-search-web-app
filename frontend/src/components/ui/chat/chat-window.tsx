import React from "react";

import { ChatMessageComponent } from "@components/ui/chat/chat-message";
import type { ChatMessage } from "@components/ui/chat/chat-message";

interface ChatWindowProps {
  messages: ChatMessage[];
  status: string;
}

export const ChatWindow: React.FC<ChatWindowProps> = ({ messages, status }) => {
  const filteredMessages = messages.filter((msg) => msg.role !== "system");

  return (
    <main className="flex-1 overflow-auto p-4 space-y-4 bg-black">
      {filteredMessages.map((msg, i) => (
        <ChatMessageComponent key={i} msg={msg} />
      ))}
      {status === "submitted" && (
        <div className="flex items-start justify-start">
          <div className="mr-2 flex-none">
            <div className="h-8 w-8 flex items-center justify-center bg-neutral-800 rounded-full text-sm font-bold">
              AI
            </div>
          </div>
          <div className="max-w-xs sm:max-w-md md:max-w-lg lg:max-w-xl px-4 py-3 rounded-2xl shadow bg-neutral-800 rounded-bl-none">
            <div className="text-sm leading-snug whitespace-pre-wrap animate-pulse text-gray-100">
              Assistant is thinking...
            </div>
          </div>
        </div>
      )}
    </main>
  );
};
