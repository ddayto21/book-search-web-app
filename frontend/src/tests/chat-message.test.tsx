/**
 * @jest-environment jsdom
 */
import React from "react";
import { render, } from "@testing-library/react";
import { ChatMessageComponent } from "../components/ui/chat/chat-message";
import type { Message as OriginalMessage } from "@ai-sdk/react";

interface Message extends OriginalMessage {
  timestamp: string;
}

describe("ChatMessageComponent", () => {
  it("renders assistant message with properly formatted markdown", () => {
    const message: Message = {
      role: "assistant",
      content: `
data: # Great choice! Historical fiction is a rich and diverse genre, full of suspense, twists, and heart-p ounding moments!
data: Here are some gripping thrill ers across different sub gen res to get you started:
data: ### Psychological Thr illers
data: 1."G one Girl"by Gill ian Flynn   - A dark, twist y tale of a marriage gone wrong, with unreliable narr ators and shocking revelations.
data: 2."The Silent Patient"by Alex Michael ides   - A psychological thriller about a woman who stops speaking after a shocking act of violence, and the therapist determined to uncover her secrets.
data: 3."Behind Closed Doors"by B. A. Paris   - A chilling story about a seemingly perfect couple with a horr ifying secret.
`,
      id: "msg-1",
      timestamp: new Date().toISOString(),
    };

    const { container } = render(<ChatMessageComponent msg={message} />);
    // Optionally, you can use a snapshot test:
    expect(container).toMatchSnapshot();

    // Or assert specific text (ensuring there are no double spaces in the rendered text)
    const renderedText = container.textContent || "";
    // Check that words like "thrillers" are correctly joined, e.g., no "thr illers"
    expect(renderedText).not.toMatch(/\b[a-zA-Z]+\s{2,}[a-zA-Z]+\b/);
  });
});
