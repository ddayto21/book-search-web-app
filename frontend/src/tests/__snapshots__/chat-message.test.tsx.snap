// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageComponent renders assistant message with properly formatted markdown 1`] = `
<div>
  <div
    class="flex items-start justify-start"
  >
    <div
      class="mr-2 flex-none"
    >
      <div
        class="flex-none mr-2"
      >
        <div
          class="h-8 w-8 rounded-full bg-gradient-to-br from-blue-600 to-purple-600 flex items-center justify-center"
        >
          <svg
            class="w-2/3 h-2/3 text-white"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4v1m6.364 1.636l-.707.707M20 12h-1M17.657 17.657l-.707-.707M12 20v-1m-6.364-1.636l.707-.707M4 12h1m2.343-5.657l.707.707M12 4a8 8 0 100 16 8 8 0 000-16z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="max-w-xs sm:max-w-md md:max-w-lg lg:max-w-xl px-4 py-3 rounded-2xl shadow bg-neutral-800 rounded-bl-none"
    >
      <div
        class="flex items-center gap-2 mb-1"
      >
        <span
          class="font-semibold text-sm text-gray-200"
        >
          LexiGPT
        </span>
        <span
          class="text-xs text-gray-400"
        >
          less than a minute ago
        </span>
      </div>
      <div
        class="text-sm leading-snug whitespace-pre-wrap text-gray-100"
      >
        <h1
          class="text-3xl font-bold mb-4 text-white"
          id="great-choice-historical-fiction-is-a-rich-and-diverse-genre-full-of-suspense-twists-and-heart-p-ounding-moments-here-are-some-gripping-thrill-ers-across-different-sub-gen-res-to-get-you-started--psychological-thr-illers-1-g-one-girlby-gill-ian-flynn--a-dark-twist-y-tale-of-a-marriage-gone-wrong-with-unreliable-narr-ators-and-shocking-revelations-2-the-silent-patientby-alex-michael-ides--a-psychological-thriller-about-a-woman-who-stops-speaking-after-a-shocking-act-of-violence-and-the-therapist-determined-to-uncover-her-secrets-3-behind-closed-doorsby-b-a-paris--a-chilling-story-about-a-seemingly-perfect-couple-with-a-horr-ifying-secret"
        >
          Great choice! Historical fiction is a rich and diverse genre, full of suspense, twists, and heart-p ounding moments! Here are some gripping thrill ers across different sub gen res to get you started: ### Psychological Thr illers 1. "G one Girl"by Gill ian Flynn -A dark, twist y tale of a marriage gone wrong, with unreliable narr ators and shocking revelations. 2. "The Silent Patient"by Alex Michael ides -A psychological thriller about a woman who stops speaking after a shocking act of violence, and the therapist determined to uncover her secrets. 3. "Behind Closed Doors"by B. A. Paris -A chilling story about a seemingly perfect couple with a horr ifying secret.
        </h1>
      </div>
    </div>
  </div>
</div>
`;
